lib/core/util.js

```diff
@@ -26,6 +26,57 @@ function isBlobLike (object) {
   )
 }
 
+function encode (val) {
+  return encodeURIComponent(val).replace(/%3A/gi, ':').replace(/%24/g, '$').replace(/%2C/gi, ',').replace(/%20/g, '+').replace(/%5B/gi, '[').replace(/%5D/gi, ']')
+}
+
+function isDate (val) {
+  return toString.call(val) === '[object Date]'
+}
+
+function isObject (val) {
+  return val !== null && typeof val === 'object'
+}
+
+function buildURL (url, queryParams) {
+  if (!queryParams) {
+    return url
+  }
+
+  const parts = []
+  for (let [key, val] of Object.entries(queryParams)) {
+    if (val === null || typeof val === 'undefined') {
+      return
+    }
+
+    if (!Array.isArray(val)) {
+      val = [val]
+    }
+
+    for (let v of val) {
+      if (isDate(v)) {
+        v = v.toISOString()
+      } else if (isObject(v)) {
+        throw new Error('Passing object as a query param is not supported, please serialize to string up-front')
+      }
+      parts.push(encode(key) + '=' + encode(v))
+    }
+  }
+
+  const serializedParams = parts.join('&')
+
+  if (serializedParams) {
+    const hashmarkIndex = url.indexOf('#')
+    if (hashmarkIndex !== -1) {
+      url = url.slice(0, hashmarkIndex)
+    }
+
+    url += (url.includes('?') ? '&' : '?') + serializedParams
```

Comment:
What happens if the same query param is twice? Should we maybe just throw if the path already includes `?`?

Reply:
Same query param being included multiple times is perfectly valid, that's a common pattern for passing arrays.
Why throw an error if we can handle this use-case with this one-liner easily? No strong opinion here, can drop if you feel strongly about it, I just see it as decreased value.

Reply:
> Same query param being included multiple times is perfectly valid, that's a common pattern for passing arrays.

I see.

> No strong opinion here, can drop if you feel strongly about it, I just see it as decreased value.

I'd prefer no surprises. Please throw for now and we can always relax it later (easier than the other direction).